pkg/erofs/erofs.go

```diff
@@ -0,0 +1,643 @@
+		if err != nil {
+			return cmd, err
+		}
+		defer os.RemoveAll(sockdir)
+		notifyPath = filepath.Join(sockdir, "notifypipe")
+		if err := syscall.Mkfifo(notifyPath, 0640); err != nil {
+			return cmd, err
+		}
+		notifyOpts = "notify_pipe=" + notifyPath
+	}
+
+	// given extractDir of path/to/some/dir[/], log to path/to/some/.dir-squashfs.log
+	extractDir = strings.TrimSuffix(extractDir, "/")
+
+	var cmdOut io.Writer
+	var err error
+
+	logf := filepath.Join(path.Dir(extractDir), "."+filepath.Base(extractDir)+"-erofsfuse.log")
+	if cmdOut, err = os.OpenFile(logf, os.O_RDWR|os.O_TRUNC|os.O_CREATE, 0644); err != nil {
+		log.Infof("Failed to open %s for write: %v", logf, err)
+		return cmd, err
+	}
+
+	fiPre, err := os.Lstat(extractDir)
+	if err != nil {
+		return cmd, errors.Wrapf(err, "Failed stat'ing %q", extractDir)
+	}
+	if fiPre.Mode()&os.ModeSymlink != 0 {
+		return cmd, errors.Errorf("Refusing to mount onto a symbolic linkd")
+	}
+
+	// It would be nice to only enable debug (or maybe to only log to file at all)
+	// if 'stacker --debug', but we do not have access to that info here.
+	// to debug erofsfuse, use "allow_other,debug"
+	optionArgs := "allow_other,debug"
+	if notifyOpts != "" {
+		optionArgs += "," + notifyOpts
+	}
+	cmd = exec.Command(erofsFuseInfo.Path, "-f", "-o", optionArgs, squashFile, extractDir)
+	cmd.Stdin = nil
+	cmd.Stdout = cmdOut
+	cmd.Stderr = cmdOut
+	cmdOut.Write([]byte(fmt.Sprintf("# %s\n", strings.Join(cmd.Args, " "))))
+	if err != nil {
+		return cmd, errors.Wrapf(err, "Failed writing to %s", logf)
+	}
+	log.Debugf("Extracting %s -> %s with %s [%s]", squashFile, extractDir, erofsFuseInfo.Path, logf)
+	err = cmd.Start()
+	if err != nil {
+		return cmd, err
+	}
+
+	// now poll/wait for one of 3 things to happen
+	// a. child process exits - if it did, then some error has occurred.
+	// b. the directory Entry is different than it was before the call
+	//    to erofsfuse.  We have to do this because we do not have another
+	//    way to know when the mount has been populated.
+	//    https://github.com/vasi/erofsfuse/issues/49
+	// c. a timeout (timeLimit) was hit
+	startTime := time.Now()
+	timeLimit := 30 * time.Second
+	alarmCh := make(chan struct{})
+	go func() {
+		_ = cmd.Wait()
+		close(alarmCh)
+	}()
+	if erofsFuseInfo.SupportsNotfiy {
+		notifyCh := make(chan byte)
+		log.Infof("%s supports notify pipe, watching %q", erofsFuseInfo.Path, notifyPath)
+		go func() {
+			f, err := os.Open(notifyPath)
+			if err != nil {
+				return
+			}
+			defer f.Close()
+			b1 := make([]byte, 1)
+			for {
+				n1, err := f.Read(b1)
+				if err != nil {
+					return
+				}
+				if err == nil && n1 >= 1 {
+					break
+				}
+			}
+			notifyCh <- b1[0]
+		}()
+
+		select {
+		case <-alarmCh:
+			cmd.Process.Kill()
+			return cmd, errors.Wrapf(err, "Gave up on erofsfuse mount of %s with %s after %s", squashFile, erofsFuseInfo.Path, timeLimit)
+		case ret := <-notifyCh:
+			if ret == 's' {
```

Comment:
what does 's' mean? Please add comment here on what we expect from notify channel and what action we're taking.